File: Week 5/1. Hamilton Cycle/HamiltonCycle.java
import java.util.Scanner;
import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;

class Graph {
    private int[][] graph;
    private int[] path;
    
    public Graph(int size) {
        this.graph = new int[size][size];
        this.path = new int[size];
    }
    
    public int[][] getGraph() {
        return graph;
    }
    
    public void setGraph(int[][] graph) {
        this.graph = graph;
    }
    
    boolean isSafe(int v, int pos) {
        if (graph[path[pos - 1]][v] == 0)
            return false;
        
        for (int i = 0; i < pos; ++i) {
            if (path[i] == v) 
                return false;
        }
        
        return true;
    }
    
    boolean hamiltonianUtil(int pos) {
        if (pos == graph.length)
            return graph[path[pos - 1]][path[0]] == 1;
            
        for (int v = 0; v < graph.length; ++v) {
            if (isSafe(v, pos)) {
                path[pos] = v;
                
                if (hamiltonianUtil(pos + 1))
                    return true;
                    
                path[pos] = -1;
            }
        }
        return false;
    }
    
    boolean isHamilton() {
        Arrays.fill(path, -1);
        
        path[0] = 0;
        return hamiltonianUtil(1);
    }
}

public class HamiltonCycle { 
    public static void main(String[] args) { 
        Scanner sc = new Scanner(System.in);
        
        int T = sc.nextInt();
        
        List<Integer> results = new ArrayList<>();
        
        for (int t = 0; t < T; ++t) {
            int n = sc.nextInt();
            int m = sc.nextInt();
            
            Graph g = new Graph(n);
            int[][] tempGraph = new int[n][n];
            
            for (int i = 0; i < m; ++i) {
                int u = sc.nextInt();
                int v = sc.nextInt();
                tempGraph[u - 1][v - 1] = 1;
                tempGraph[v - 1][u - 1] = 1;
            }
            g.setGraph(tempGraph);
            
            results.add(g.isHamilton() ? 1 : 0);
        }
        
        for (int result : results) {
            System.out.println(result);
        }
        
        sc.close();
    }
}

File: Week 5/1. Hamilton Cycle/hamilton_cycle.py
def is_safe(v, graph, path, pos):
    if graph[path[pos - 1]][v] == 0:
        return False
        
    if v in path:
        return False
    return True
    
def hamilton_util(graph, path, pos):
    if pos == len(graph):
        return graph[path[pos - 1]][path[0]] == 1 
        
    for v in range(1, len(graph)):
        if is_safe(v, graph, path, pos):
            path[pos] = v
            
            if hamilton_util(graph, path, pos + 1):
                return True
                
            path[pos] = -1
            
    return False
        
def is_hamilton(graph):
    path = [0] * len(graph)
    path[0] = 0
    
    if not hamilton_util(graph, path, 1):
        return 0
        
    return 1 
    
if __name__ == "__main__":
    T = int(input())
    results = []
    
    for _ in range(T):
        n, m = map(int, input().split())
        graph = [[0 for _ in range(n)] for _ in range(n)]
        
        for _ in range(m):
            u, v = map(int, input().split())
            graph[u-1][v-1] = 1
            graph[v-1][u-1] = 1  

        results.append(is_hamilton(graph))
        
    for result in results:
        print(result)
        
